drivers/serial/uart_b91.c

```diff
@@ -518,6 +522,36 @@ static void uart_b91_irq_callback_set(const struct device *dev,
 
 #endif /* CONFIG_UART_INTERRUPT_DRIVEN */
 
+#ifdef CONFIG_PM_DEVICE
+
+static int uart_b91_pm_action(const struct device *dev, enum pm_device_action action)
+{
+	volatile struct uart_b91_t *uart = GET_UART(dev);
+	struct uart_b91_data *data = dev->data;
+
+	switch (action) {
+	case PM_DEVICE_ACTION_RESUME:
+		/* reset TX/RX byte index */
+		data->tx_byte_index = 0;
+		data->rx_byte_index = 0;
+		uart->status |= UART_RX_RESET_BIT | UART_TX_RESET_BIT;
+		break;
+
+	case PM_DEVICE_ACTION_SUSPEND:
+		/* wait for TX done before entering suspend mode */
+		while (!(uart->txrx_status & UART_TXRX_STATUS_TX_DONE)) {
+		};
```

Comment:
please make use of `pm_policy_state_lock_get`/`pm_policy_state_lock_put` API, the PM callback is not a place where one should run busy loops. That means system is trying to enter a PM state that is not supported when UART is active.

Reply:
Entering suspend on Telink B91 affects all peripheral devices
It's not just core functionality as most other vendors.
It mean no way to separate CPU sleep and peripheral devices sleep. 
So if some device is busy with its operation it should prevent entering
suspend mode at all.
from zephyr/subsys/pm/pm.c
I see the only way to prevent enter suspend
is to use pm_suspend_devices();
in that case corresponding device should return error
or finish its operations (as it's implemented).
Returning an error produces kernel trace about no suspend entered.
So it's implemented by this way.
Suggested CONFIG_PM_DEVICE_RUNTIME disables preventing core suspend entering.
May be I missed something. If there is a better option please suggest it.

Reply:
polling is removed from PM callback.